pkg/tkctl/cmd/upinfo/upinfo.go

```diff
@@ -0,0 +1,213 @@
+	o := NewUpInfoOptions(streams)
+
+	cmd := &cobra.Command{
+		Use:     "upinfo",
+		Short:   "Show tidb upgrade info.",
+		Example: upinfoExample,
+		Long:    upinfoLongDesc,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(o.Complete(tkcContext, cmd, args))
+			cmdutil.CheckErr(o.Run())
+		},
+		SuggestFor: []string{"updateinfo", "upgradeinfo"},
+	}
+
+	return cmd
+}
+
+func (o *UpInfoOptions) Complete(tkcContext *config.TkcContext, cmd *cobra.Command, args []string) error {
+
+	clientConfig, err := tkcContext.ToTkcClientConfig()
+	if err != nil {
+		return err
+	}
+
+	if tidbClusterName, ok := clientConfig.TidbClusterName(); ok {
+		o.TidbClusterName = tidbClusterName
+	} else {
+		return cmdutil.UsageErrorf(cmd, infoUsage)
+	}
+
+	namespace, _, err := clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+	o.Namespace = namespace
+
+	restConfig, err := clientConfig.RestConfig()
+	if err != nil {
+		return err
+	}
+	tcCli, err := versioned.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+	o.TcCli = tcCli
+	kubeCli, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+	o.KubeCli = kubeCli
+
+	return nil
+}
+
+func (o *UpInfoOptions) Run() error {
+
+	tc, err := o.TcCli.PingcapV1alpha1().
+		TidbClusters(o.Namespace).
+		Get(o.TidbClusterName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	setName := controller.TiDBMemberName(tc.Name)
+	set, err := o.KubeCli.AppsV1beta1().StatefulSets(o.Namespace).Get(setName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	podList, err := o.KubeCli.CoreV1().Pods(o.Namespace).List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s,%s=%s", label.InstanceLabelKey, tc.Name, label.ComponentLabelKey, "tidb"),
+	})
+	if err != nil {
+		return err
+	}
+	msg, err := renderTCUpgradeInfo(tc, set, podList)
+	if err != nil {
+		return err
+	}
+	fmt.Fprint(o.Out, msg)
+	return nil
+}
+
+func renderTCUpgradeInfo(tc *v1alpha1.TidbCluster, set *apps.StatefulSet, podList *v1.PodList) (string, error) {
+	return readable.TabbedString(func(out io.Writer) error {
+		w := readable.NewPrefixWriter(out)
+		dbPhase := tc.Status.TiDB.Phase
+		w.WriteLine(readable.LEVEL_0, "Name:\t%s", tc.Name)
+		w.WriteLine(readable.LEVEL_0, "Namespace:\t%s", tc.Namespace)
+		w.WriteLine(readable.LEVEL_0, "CreationTimestamp:\t%s", tc.CreationTimestamp)
+		w.WriteLine(readable.LEVEL_0, "Statu:\t%s", dbPhase)
+		if dbPhase == v1alpha1.UpgradePhase {
+			if len(podList.Items) != 0 {
+				pod := podList.Items[0]
+				w.WriteLine(readable.LEVEL_0, "Image:\t%s ---> %s", pod.Spec.Containers[0].Image, tc.Spec.TiDB.Image)
+			}
+		}
+		{
+			w.WriteLine(readable.LEVEL_1, "Name\tState\t")
+			w.WriteLine(readable.LEVEL_1, "----\t-----\t")
+			{
+				updateReplicas := set.Spec.UpdateStrategy.RollingUpdate.Partition
+
+				if len(podList.Items) != 0 {
+					for _, pod := range podList.Items {
+						var state string
```

Comment:
The logic inside the `for` loop is nested deeply, how about abstract this as a function?

Reply:
It is not complicate so I think it is not necessary。